### DISCUSSION TITLE
Sometimes queries with regex are slow

### DISCUSSION THREAD

**Discussion by BBQigniter** [prometheus-operator/prometheus-operator#5321] [General]
Hi,

we have a very strange problem with some prometheus instances setup via prometheus-operator on some of our clusters. 

I tried with latest and older prometheus versions but we always or very often have the problem that queries especially with not too complex regex in labels are in some cases super-slow (for example Grafana is doing queries like that and often timing out in our case). But it seems random and it seems not to matter if we run prometheus with 1 or 2 replicas.

For example I have here this query exectued via prometheus' web-ui:

```
kubelet_containers_per_pod_count_sum{job="kubelet",instance=~"(192.168.123.47:10250|192.168.123.48:10250|192.168.123.49:10250)"}
```

which takes about 180ms

Then I execute the query slightly different showing the same results:

```
kubelet_containers_per_pod_count_sum{job="kubelet",node=~"(dc-lab-vnode-001|dc-lab-vnode-002|dc-lab-vnode-003)"}
```

which takes whopping ~1000-2000+ms - *BUT* sometimes it also can be very fast if you are lucky. I absolutely cannot find any pattern when or why some queries take so long. First I thought it has something to do with services monitored via "ServiceMonitor"-manifests - but I saw slowness also with normal prometheus-jobs.

Similar queries on other prometheus instances that are not set up via prometheus-operator do not have this problem. All our prometheus instances remote-read and -write to an influxdb, sometimes using the same influxdb-node but a separate database with pretty much the same configuration (only retention may vary).

I also have setup an additional prometheus instance old-style like without the usage of prometheus-operator and only 1 replica. The instance is also configured with remote-read and -write to an influxdb that has even much less computing resources and there the queries are faster and I haven't seen similar issues.

Also checked cardinality of the `instance` and `node` label - via `count (group by (node)(kubelet_containers_per_pod_count_sum))` but they are negligible resulting with a value of 4-6. Also the label-values length cannot be the problem like suggested in another issue I found here in the issues.

Is there probably anyone who had/has similar problems and knows how to fix this? 

Thanks in advance for any hint :)

**Comment by simonpasquier** [comment]
I'm not sure that it's an issue with the Prometheus operator. Have you tried retrieving Go profiles to understand where the time is spent?

**Reply by BBQigniter** [reply]
hmm, interesting. never heard of this. will check this out and read into this topic as soon as possible when I have time.

**Reply by BBQigniter** [reply]
ok, I now had a brief look at how to do "go profiling" - but as I have not really a clue about go and debugging some go-application, I hope someone can give me some hints what to look for. Also my approach is probably a little bit naive - but well, I don't know it better for the moment.

first I tried

```
go tool pprof -svg http://the-affected-prometheus.example.com/debug/pprof/profile?seconds=10 > profile.svg
```

and while this was running I executed a query that "thankfully" ran for about 3-4 seconds. But in the nifty svg-file I couldn't see anything obvious that would explain why or where the query took so long.

then I tried

```
curl -o trace.out  http://the-affected-prometheus.example.com/debug/pprof/trace?seconds=10
```

again, while the trace was being created, I executed the query. After the trace was completed I started

```
go tool trace trace.out
```

and it opened my default-browser (in my case Firefox*). But what's shown after you click the links on the start-page, is just overwhelming :open_mouth: - what should I look for??? In the "Trace View" again there was nothing obvious for me. I imagined that there will be somewhere some bar or so with the length corresponding to the time the query ran, from where on you could dig into the details.

How could I limit the trace/profile/whatever to show only what's going on from the time on I pressed "Execute" in the Prometheus-WebUI?

This is quite like rocket-science for me :grimacing:

*There is some warning showing that I should start the browser with `--enable-blink-features=ShadowDOMV0,CustomElementsV0,HTMLImports`. But if I do that, the browser complains about an unknown flag. Also tried with an extra installed google-chrome, but seeing same issue. So I even do not know if the "Trace View" is working correctly :shrug:

**Reply by simonpasquier** [reply]
Please run and share the cpu.out file
`
curl -o cpu.out  http://the-affected-prometheus.example.com/debug/pprof/profile?seconds=10 
`
Doing the same on a Prometheus that doesn't exhibit the same issue would be great too.

**Reply by BBQigniter** [reply]
@simonpasquier, thanks for your patience and help

I think I know why I didn't see before anything (at least obvious for me) in the trace. As our prometheus-operator managed instances are running with 2 replicas, I think the first trace was taken from the other instance where the query was not running.

So I had to add the `INGRESSCOOKIE` to the `curl`-command and then I spotted in the trace's "Go routine analysis", in the routine with `mergeGenericQuerier` in the name, timings that correspond to the time taken for the query to finish.

Also interesting was, today in the morning the same queries were blazing fast. But after I switched in the prometheus-ui to the graph and set the range to more than 12h, the query time took again 1-2+ seconds. As far as I have understood, the prometheus pods keep a few hours in their local wal-files inside, don't they? I restarted now the pods and "surprise" the queries are super-slow again. I now again created a trace-profile and a cpu-profile like you requested. A quick look into the trace, showed again the long `mergeGenericQuerier` timings.

I did the same with a "old-style" setup prometheus pod which is running only with 1 replica. There the queries do not take that long and stay below 1s no matter how large I set the range for the graph.

As far as I have understood from the documentations of prometheus-operator, the prometheus instances are setup with a config that shards the metrics, so that prometheus-instance-1 and prometheus-instance-2 do not create duplicated metrics, right? Is then my assumption correct, that this `mergeGenericQuerier`-routine sends the query to the datastore and then merges the metrics somehow so that they can be shown correctly in the UI, if there is not enough data in the local wal-files? With more complicated regexs this could explain why it's taking longer, as the merge-algorithm has to maybe take a little bit more into account to show the correct data?

That's only my personal assumption, sry if I'm writing BS :neutral_face:

Hope those profiles can help somehow.

[prometheus-pprofs.zip](https://github.com/prometheus-operator/prometheus-operator/files/10764980/prometheus-pprofs.zip)

**Comment by simonpasquier** [comment]
Hmm the "managed" profile shows that most of the CPU time is spent by remote-read requests while these aren't visible in the legacy Prometheus. Can you share the full config for both installations? I'd like to make sure that we're not comparing oranges to apples.

**Reply by BBQigniter** [reply]
> ...oranges to apples

well it may be a kind of comparing "Red Delicious" to "Golden Delicious". Both are apples - but slightly different 😉.

I also checked yesterday, if it may be some strange DNS issue in the pod. 

* I created the DNS-entry for the remote-read/write destination similarily like we do for our "legacy-setup" 
  ```
  kind: Service
  apiVersion: v1
  metadata:
    name: influxdb
    namespace: monitor
  spec:
    type: ExternalName
    externalName: some-influxdb.our.domain
  ```
* I created the DNS-entry as `/etc/hosts`-entry via `hostAliases`
* I changed the remote-read/write destination to use the IP instead of FQDN

But queries with regex other than ".*" in the label were always strangely-slow on the operator-managed instances.

The InfluxDB-hosts hold also other databases for other Prometheus-instances, the hosts do not have any CPU/network/IO-issues as far as I have seen. All our "legacy-setups" do not have the slow-query issues. The InfluxDB version is 1.8.6.

If you want the whole config, including the Kubernetes-manifests, this will take a while for me, to obfuscated needed stuff, the whole setup was done with kustomize by a colleague who left the company weeks ago. Attached you find the Prometheus-configurations copied via the web-UI.

In the "managed-prometheus-config" we use parts from the legacy-config in the "additional-scrape-config". In the old-configs we supply the credentials to the database in the URI, in the new config via "basic-auth"-option (I haven't yet tried to change that)

[prometheus-configs.zip](https://github.com/prometheus-operator/prometheus-operator/files/10791229/prometheus-configs.zip)
